relays/lib-substrate-relay/src/finality_base/engine.rs

```diff
@@ -319,4 +341,82 @@ impl<C: ChainWithGrandpa> Engine<C> for Grandpa<C> {
 			operating_mode: BasicOperatingMode::Normal,
 		})
 	}
+
+	async fn finality_verification_context<TargetChain: Chain>(
+		target_client: &impl Client<TargetChain>,
+		at: HashOf<TargetChain>,
+	) -> Result<Self::FinalityVerificationContext, SubstrateError> {
+		let current_authority_set_key = bp_header_chain::storage_keys::current_authority_set_key(
+			C::WITH_CHAIN_GRANDPA_PALLET_NAME,
+		);
+		let authority_set: AuthoritySet = target_client
+			.storage_value(at, current_authority_set_key)
+			.await?
+			.map(Ok)
+			.unwrap_or(Err(SubstrateError::Custom(format!(
+				"{} `CurrentAuthoritySet` is missing from the {} storage",
+				C::NAME,
+				TargetChain::NAME,
+			))))?;
+
+		authority_set.try_into().map_err(|e| {
+			SubstrateError::Custom(format!(
+				"{} `CurrentAuthoritySet` from the {} storage is invalid: {e:?}",
+				C::NAME,
+				TargetChain::NAME,
+			))
+		})
+	}
+
+	async fn synced_headers_finality_info<TargetChain: Chain>(
+		target_client: &impl Client<TargetChain>,
+		at: TargetChain::Hash,
+	) -> Result<Vec<HeaderGrandpaInfo<HeaderOf<C>>>, SubstrateError> {
+		let stored_headers_grandpa_info: Vec<StoredHeaderGrandpaInfo<HeaderOf<C>>> = target_client
+			.state_call(at, C::SYNCED_HEADERS_GRANDPA_INFO_METHOD.to_string(), ())
+			.await?;
+
+		let mut headers_grandpa_info = vec![];
+		for stored_header_grandpa_info in stored_headers_grandpa_info {
+			headers_grandpa_info.push(stored_header_grandpa_info.try_into().map_err(|e| {
+				SubstrateError::Custom(format!(
+					"{} `AuthoritySet` synced to {} is invalid: {e:?} ",
+					C::NAME,
+					TargetChain::NAME,
+				))
+			})?);
+		}
+
+		Ok(headers_grandpa_info)
+	}
+
+	async fn generate_source_key_ownership_proof(
+		source_client: &impl Client<C>,
+		at: C::Hash,
+		equivocation: &Self::EquivocationProof,
+	) -> Result<Self::KeyOwnerProof, SubstrateError> {
+		let set_id = equivocation.set_id();
+		let offender = equivocation.offender();
+
+		let opaque_key_owner_proof = source_client
+			.generate_grandpa_key_ownership_proof(at, set_id, offender.clone())
```

Comment:
So far our relay has been able to work with non-archive nodes. It has never accessed ancient blocks (ancient block by default is something near `best - 200` or so). Do we plan to call this method for ancient blocks now? If so, it'll change requirement for nodes and e.g. we won't be ably to run equivocation detection using public RPC nodes.

Reply:
I wonder how the regular equivocation reports work? I mean - on regular nodes? Do they assume that it is called at archive node?

Reply:
> I think the grandpa client only checks for equivocations in the current round or recent rounds. So not being able to access ancient blocks shouldn't be a problem.

But are you planning to store collected justifications somewhere and check them on restart (e.g. if relayer is crashed or there were e.g. connection-with-the-target-node issue)? Then you'll probably need to access the ancient block state - to read events + to get equivocation proofs.

I don't know about historical pallet, but I doubt it'll help with reading events from ancient blocks. So probably we need to think of running this agent on nodes that are storing states for at least `<unbonding-period>` of blocks?

Reply:
I'm not sure, but not in the initial equivocations detector version anyway. I would add this in a future PR after we have a basic equivocations detector. So I didn't plan that far. 

I'm not completely sure about this, because if the relayer is offline, we're also losing ephemeral finality proofs from the source as well, and even if we get the proofs that were synced at the target, we might not be able to look for equivocations. It might help in some situations. But anyway if the relayer was offline for more then 200 target blocks, we lose quite a lot of source ephemeral proofs.

> I don't know about historical pallet, but I doubt it'll help with reading events from ancient blocks. So probably we need to think of running this agent on nodes that are storing states for at least <unbonding-period> of blocks?

Yes, the historical pallet can only help us with key ownership proofs at best. If we're lucky it can generate key ownership proofs for older validator sets.

I don't know if we need to get older target blocks. If we get older target blocks, we probably won't have the ephemeral source proofs corresponding to them anyway. I was thinking at least for the initial version to have the relayer check for equivocations only in the newly generated blocks at target. If there are multiple relayers running, this should be ok. Even if at times 1 relayer fails or restarts.

> But are you planning to store collected justifications somewhere and check them on restart (e.g. if relayer is crashed or there were e.g. connection-with-the-target-node issue)?

I'm not completely sure about saving the state either, for the same reason. If the relayer is offline, we're also losing ephemeral finality proofs from the source as well, and even if we get the proofs that were synced at the target, we might not be able to look for equivocations. It might help in some situations. But anyway if the relayer was offline for more then 200 target blocks, we lose quite a lot of source ephemeral proofs.

Reply:
It is true that ephemeral justifications will be lost if we'll lose connection to the source node. But we can collect them and use later if this connection is live but we e.g. have lost connection with the target node.

Ok, I thought that you're planning to impl that persistent storage in the initial version. Without it my comment is not actual anymore :)